src/lib.rs

```diff
@@ -39,6 +41,96 @@ impl Default for ConstraintList {
 }
 
+struct ListFilter {
+    constraints: Rc<RefCell<Vec<Vec<i32>>>>,
+    length_filter: HashSet<usize>,
+    cell_filter: HashSet<usize>,
+    cell_constraints: HashMap<(i32, i32), HashSet<usize>>,
+}
+
+impl ListFilter {
+    pub fn new(constraints: Rc<RefCell<Vec<Vec<i32>>>>) -> Self {
+        Self {
+            constraints: Rc::clone(&constraints),
+            length_filter: (0..constraints.borrow().len()).collect(),
+            cell_filter: (0..constraints.borrow().len()).collect(),
+            cell_constraints: HashMap::new(),
+        }
+    }
+
+    fn apply_filters(&self) -> Vec<Vec<i32>> {
+        let mut final_set = self.length_filter.clone();
+
+        // Add additional filters with && in the same closure
+        final_set.retain(|index| self.cell_filter.contains(index));
+
+        let mut final_list = Vec::new();
+        for index in final_set {
+            final_list.push(self.constraints.borrow()[index].clone());
+        }
+
+        final_list
+    }
+
+    pub fn reinit(&mut self) {
+        self.create_cell_map();
+        self.clear_all();
+    }
+
+    fn create_cell_map(&mut self) {
+        for row in 1..=9 {
+            for col in 1..=9 {
+                self.cell_constraints.insert((row, col), HashSet::new());
+            }
+        }
+        for (index, list) in self.constraints.borrow().iter().enumerate() {
+            for identifier in list {
+                let (row, col, _) = identifier_to_tuple(*identifier);
+                if let Some(cell_set) = self.cell_constraints.get_mut(&(row, col)) {
+                    cell_set.insert(index);
+                }
+            }
+        }
+    }
+
+    /// Filters the constraints by the given max_length.
+    pub fn by_max_length(&mut self, max_length: i32) -> Vec<Vec<i32>> {
+        let mut filter_set = HashSet::new();
+        for (index, constraint) in self.constraints.borrow().iter().enumerate() {
+            if constraint.len() as i32 <= max_length {
+                filter_set.insert(index);
+            }
+        }
+        self.length_filter = filter_set;
+        // Return new filtered list
+        self.apply_filters()
+    }
+
+    pub fn by_cell(&mut self, row: i32, col: i32) -> Vec<Vec<i32>> {
+        if let Some(cell_set) = self.cell_constraints.get(&(row, col)) {
+            self.cell_filter = cell_set.clone()
+        }
+        self.apply_filters()
+    }
+
+    pub fn clear_length(&mut self) -> Vec<Vec<i32>> {
+        self.length_filter = (0..self.constraints.borrow().len()).collect();
+        self.apply_filters()
+    }
+
+    pub fn clear_cell(&mut self) -> Vec<Vec<i32>> {
+        self.cell_filter = (0..self.constraints.borrow().len()).collect();
+        self.apply_filters()
+    }
+
+    pub fn clear_all(&mut self) -> Vec<Vec<i32>> {
+        // This will apply the filters several times, but avoids code repetition
+        // Change if this becomes a problem!
+        self.clear_length();
+        self.clear_cell()
+    }
```

Comment:
This is a bit weird when apply, functions that set max length and cell and clear functions all return the filtered constraints. Shouldn't there be one method that returns the constraints and all the others would just modify the internal state.

Reply:
This was done to make the interface as simple as possible for the GUI code. It also reduces the risk of accidentally changing the internal filter state without actually applying the changes to the GUI. This was a topic of some discussion for the group working on this, so there's pros and cons to both. We did keep the apply function as private, so that no one will thing that it needs to be called separately.

Reply:
So a simple interface in my mind is one where you can do a particular thing only in one way. Currently there is no way to just get the resulting constraints without modifying the internal state but there are multiple functions that both modify and return the constraints. Also function reinit() only modifies the internal state and doesn't return the resulting constraints after. Regarding the risk of not applying the changes to GUI, I think that having only a one endpoint to get the constraints would reduce that risk because then it would be clear by only looking at the method signatures that what function is to be called to get the filtered constraints. Also reinit() at its current form would infact cause the risk that you are refering to because it's the only public function that doesn't return.

Reply:
You make a good point RE: reinit(). As for the rest, I don't know that I *necessarily* agree. We don't *want* anyone to access the constraints through the ListFilter, except when the list changes; that's what rendered_constraints in the App struct is for. I'll leave the rest of this discussion and the final decision to those who are able to actually actively work on the project  right now :D